### ISSUE TITLE
How do we enable custom fonts?

### ISSUE THREAD

**Issue by davidlimjh** [Automattic/node-canvas#761]
I saw the example.js for custom fonts and tried to use it but it seems I dont have the canvas.font dependency. Im running the node-canvas installed using npm.

Is it possible for me to use custom fonts by simply installing them on my Windows installation then specifying the font?

**Comment by chearon** [comment]
For some reason there isn't FreeType or Pango support for the Windows build but it is possible. I have it working in #715. When that gets merged you should have no problem using custom fonts. It's nearing completion but there will be a lot of review and testing to do after that.

If editing the source and rebuilding will work for you as a temporary solution I can post a `binding.gyp` that should work for you

**Comment by davidlimjh** [comment]
That would work for me. Thank you very much!

**Comment by FZFalzar** [comment]
Hi @chearon , regarding this I would also like to know about how the Windows build handles fallback fonts since they don't seem to use Windows' own mechanism. 

As far as I've researched, the Pango backend should be able to take in a "pango.aliases" file. 
However, placing it in the directory like C:/GTK/etc/pango/pango.aliases does not seem to have any effect, nor does editing the gyp script to copy those into the project directory for canvas

Some insight is greatly appreciated!

**Comment by chearon** [comment]
I know it does font selection (finding the closest weight, style, etc) in Pango space, so yes that is going to be different than Windows does it.

Are you trying to specify the font it should use if the one you specify is not found? In that case it would use "sans", so if you point that rule to an alternate font name in the pango.aliases it might work

**Comment by FZFalzar** [comment]
Yes I am trying to specify fallback fonts, however the problem lies in figuring out where the aliases file goes since the way node-canvas ships has gtk required as an external installation rather than being shipped together as libs/dlls with the program (like most windows apps built with gtk )

**Comment by chearon** [comment]
OK, I think you are putting it in the right place... maybe the version of GTK that is linked to in the wiki doesn't have that support built in. You could try getting the DLLs from mingw which is the current recommended way. I'll try to mess around with it when I get a second

**Comment by chearon** [comment]
@FZFalzar bad news, Pango doesn't read that file anymore: 

https://mail.gnome.org/archives/commits-list/2015-April/msg00796.html

**Comment by FZFalzar** [comment]
Well crud, if that's the case, is it possible to register fonts of a similar family together (lets say Noto Sans and its language variants) under a single user-defined font family (not limited by the font's own family name), and then by calling ctx.font = "userdefinedfamilyname" all the fallbacks and missing glyphs can be handled?

**Comment by chearon** [comment]
Have you tried #715? If you clone my repo, checkout `prefer-pango` and `npm install` it should do all of that. I did notice one issue in the Windows build in helping you so I'll push that up now

**Comment by FZFalzar** [comment]
As a matter of fact yes I did, how I am currently handling my fonts within my nodejs application goes like this:
`
Canvas.registerFont(this.mont_bold_font_dir, {
      family: 'Montserrat'
    })
    Canvas.registerFont(this.nsthai_font_dir, {
      family: 'Noto Sans Thai'
    })
    Canvas.registerFont(this.nscjk_reg_font_dir, {
      family: 'Noto Sans CJK JP Regular'
    })
`
Somewhere in drawing code:
`ctx.font = '11pt \'Noto Sans CJK JP Bold\''`

This draws fine, but as a result certain fonts from other language variants will not be drawn
If I specify the partial name of the font like just "Noto Sans", then it'll give a Pango warning and fallback to Sans

Also when I checked out your repo on the latest commit on prefer-pango, I was unable to build successfully with Win64 GTK version 2.22 off GNOME with the error being g_list_free_full within register_font.cc, which I'm guessing is due to an outdated glib. Which version of GTK are you currently building with?

**Comment by chearon** [comment]
Yeah, I just pushed the change to fix that! It will build now. The Windows GTK that is on the wiki is a bit behind. If you weren't able to build it, how would your calls to `Canvas.registerFont` have worked? 

You can't specify "partial" family names, Pango will look up the font by name with a lower case comparison, if it does not find it it will select Sans. I don't know if there is a way for Pango to let you to specify which font to use if a font doesn't have all the characters in a run of text - that sounds like something more for FontConfig and I haven't tried getting that working on Windows

**Comment by FZFalzar** [comment]
Yup it builds fine now, if you could get fontconfig working or figure out how to explicitly tell Pango which fonts to use then that's a plus :) Otherwise, thanks for your insight, it looks like I'll have to live with broken glyphs for now or figure out a way via unicode subrange detection

**Comment by chearon** [comment]
Still thinking about this lol... did you try comma-separating the families you want when you set `ctx.font`? It seems like you might just want `ctx.font = "11pt Montserrat, 'Noto Sans Thai', 'Noto Sans CJK JP Regular'";`, which should absolutely work. Your run of text should get all 3 fonts and Pango would be able to pick the right one based on the language I *think*

**Comment by FZFalzar** [comment]
Just tried that, unfortunately didn't work haha, but I did notice these:
1. No Pango warnings, the font seemed to render normally
2. It just picks the first font from the left and renders that, missing glyphs come out as unknown symbols (white box things). Varying which font went first just renders the text based on that font
3. Putting the font size in front of one, or several of the fonts specified seems to have no effect, except only for the first font in the list

**Comment by chearon** [comment]
yeah, I tried it too, it's because node-canvas just picks the first font in the list. Pango supports a list of families though so I think this is possible and I'm almost done with a crude implementation.

[You can't specify multiple sizes/weights/etc in a `font` rule](https://developer.mozilla.org/en-US/docs/Web/CSS/font), if you think about it that wouldn't make sense anyways though! Just comma separated family names at the end of the list

FWIW you should never get unknown glyph symbols printed if you have fallback fonts installed on your system

**Comment by chearon** [comment]
OH i just realized pango fallback fonts aren't working in Windows

My god, there are so many unfinished details on OS X and windows

**Comment by chearon** [comment]
I got font fallbacks from the font list working in a very crude way on Windows. It's funny, if you only specify the Thai font it can't even fallback to Roman characters. Maybe they fixed this in later versions of Pango on Windows...

![working](http://i.imgur.com/b8DDYbE.png)

It's on my branch here https://github.com/chearon/node-canvas/tree/multiple-families-crude

I'll get this added properly to #715 soon. Thanks for reporting this and working with me on it!!

**Comment by FZFalzar** [comment]
Just tested your latest commit, it seems to be working as intended except that registering same fonts but with different weights under the same original family name it breaks.
`
  Canvas.registerFont(this.mont_bold_font_dir, {
      family: 'Montserrat',
      weight: 'bold'
    })
    Canvas.registerFont(this.mont_reg_font_dir, {
      family: 'Montserrat',
      weight: 'regular'
    })
`
In my console I see font family name being merged and repeated like `MontserratMontserrat`, whereas specifying the family as A,B,C:
`
    Canvas.registerFont(this.nscjk_reg_font_dir, {
	family: 'A'
    })
    Canvas.registerFont(this.nst_reg_font_dir, {
	family: 'B'
    })
    Canvas.registerFont(this.mont_reg_font_dir, {
        family: 'D'
    })
    Canvas.registerFont(this.ns_reg_font_dir, {
	family: 'C'
    })
`
 seems to be replaced and separated properly with commas like `Montserrat, Noto Sans CJK JP`

**Comment by chearon** [comment]
OK try now (you'll have to fetch and `git reset --hard origin/multiple-families-crude`, sorry). I didn't have that part completed. I think it's almost ready to add to #715 let me know if you have any issues with it

**Comment by FZFalzar** [comment]
I seem to be getting some odd warning when I try to specify weight:
`Canvas.registerFont('fonts/NotoSansCJKjp-Bold.otf', {
      //family: 'Noto Sans CJK JP'
	  family: 'E',
	  weight: 'bold'
    })
ctx.font = "7pt bold E"`

The text still renders as intended, but however this appears in the console:
`Pango-CRITICAL **: pango_font_description_to_string: assertion `desc != NULL' failed`

The same also happens when I do not specify weight in registerFont, but then try to use it within ctx.font; the text renders and spouts the same warning 

Oddly enough, when I tried it with Montserrat, there were no warnings

**Comment by chearon** [comment]
@FZFalzar try again, I was not resolving font names that weren't custom fonts properly and I think it could have been related to that. I think it is now pretty much ready to add to #715

**Comment by arthurvanl** [comment]
Any updates on this problem? I still can't use fonts

```sh
Pango-WARNING **: 07:12:35.291: couldn't load font "Custom-Font Not-Rotated 12px", falling back to "Sans Not-Rotated 12px", expect ugly output```

**Comment by bvitlas** [comment]
> Any updates on this problem? I still can't use custom fonts
> 
> ```shell
> Pango-WARNING **: 07:12:35.291: couldn't load font "Custom-Font Not-Rotated 12px", falling back to "Sans Not-Rotated 12px", expect ugly output```
> ```

did you solve this out ? same here actually :)
